src/handlers/mod.rs

```diff
@@ -48,6 +53,31 @@ impl SeatHandler for State {
 delegate_seat!(State);
 delegate_tablet_manager!(State);
 delegate_pointer_gestures!(State);
+delegate_text_input_manager!(State);
+
+impl InputMethodHandler for State {
+    fn new_popup(&mut self, surface: PopupSurface) {
+        if let Err(err) = self.niri.popups.track_popup(PopupKind::from(surface)) {
+            warn!("error tracking ime popup {err:?}");
+        }
+    }
+    fn parent_geometry(&self, parent: &WlSurface) -> Rectangle<i32, Logical> {
+        self.niri
+            .global_space
+            .elements()
+            .find_map(|window| {
```

Comment:
So, the global space doesn't actually have any windows, it's just for the outputs. It should be something like:
```
self.niri.monitor_set.find_window_and_output(parent).map(|window, _| window.geometry())
```